### PR TITLE
refactor: switch to event-based undo

### PR DESCRIPTION
**Pull request by PurpleMyst**
this is preliminary to #30

### CODE REVIEW

File: js/canvasmanager/events.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  }
+}
+export class ClearEvent implements Event {
+  constructor(private objects: DrawObject[]) { }
+
+  undo(cm: CanvasManager): void {
+    console.assert(cm.objects.length === 0, "undoing clear without empty objects");
+    cm.objects = this.objects;
+  }
+}
```
Comment (Akuli) [inline js/canvasmanager/events.ts:19]: Why is this simply not a part of canvasmanager.js? It's not java. There's no need to put everything into many tiny files.

File: tsconfig.json [hunk 2]
```diff
@@ -1,7 +1,8 @@
 {
   "compilerOptions": {
     "module": "esnext",
-    "target": "esnext",
+    "target": "es2019",
+    "lib": ["ES2019", "DOM"],
```
Comment (Akuli) [inline tsconfig.json:5]: Why has this worked before without `"DOM"` in `"lib"`, and why do we need to add `"DOM"` now?

**Review by Akuli — CHANGES_REQUESTED** [review]
Looks good! I would like some small things to be changed.

File: tsconfig.json (hunk 2 above)
Comment (PurpleMyst) [reply tsconfig.json:5]: We didn't have a "lib" key at all before. This now specifies the libraries that TypeScript knows we have, so ES2019's and the DOM's. I'm not sure if it's necessary, but the official TypeScript page mentions it (see commit message)

File: js/canvasmanager/events.ts (hunk 1 above)
Comment (PurpleMyst) [reply js/canvasmanager/events.ts:19]: There's like 3 events and it felt like it was getting in the way of the main canvas manager code to me 🤷🏻‍♂️

Comment (Akuli) [reply js/canvasmanager/events.ts:19]: canvasmanager.ts would easily be short enough even with the events in it. I'd rather have it in one file. If it **really** needs to be in a separate file, then the directory structure should be e.g. `canvasmanager-events.ts`, `canvasmanager.ts` and `drawobjects/pen.ts`.

Comment (PurpleMyst) [reply js/canvasmanager/events.ts:19]: I think having the events in the same file is kinda noisy, and I don't really see the point of naming something with a prefix when you could just as easily use a directory tbh

But if you want I can merge events.ts and canvasmanager.ts

Comment (Akuli) [reply js/canvasmanager/events.ts:19]: That would be nicer than prefixing or many nested directories

**Comment by PurpleMyst** [comment]
By the way, adding in redo to this system is practically free. I could do it if you're fine with it.

**Comment by Akuli** [comment]
What do you think? (turns out that I can't request a review from you because you created the PR)

**Review by Akuli — APPROVED** [review]

**Comment by PurpleMyst** [comment]
LGTM :>

**Pull request merged**